software/contrib/egressus_melodiam.py

```diff
@@ -0,0 +1,969 @@
+        """Update the screen only if something has changed. oled.show() hogs the processor and causes latency."""
+
+        # Only update if something has changed
+        if not self.screenRefreshNeeded:
+            return
+        # Clear screen
+        oled.fill(0)
+
+        # Selected output
+        oled.fill_rect(108, 0, 20, 9, 1)
+        oled.text(f"{self.selectedOutput + 1}", 115, 1, 0)
+
+        # Show division for selected output
+        number = self.outputDivisions[self.selectedOutput]
+        x = 111 if number >= 10 else 115
+        oled.text(f"{number}", x, 12, 1)
+
+        # Show wave for selected output
+        self.drawWave()
+
+        if self.unClockedMode:
+
+            if self.averageMsBetweenClocks != 0:
+                oled.text(f"{int(self.averageMsBetweenClocks * 2)} ms", 31, 1, 1)
+
+        else:
+
+            # Draw pattern length
+            row1 = ""
+            row2 = ""
+            row3 = ""
+            row4 = ""
+            if self.patternLength > 24:
+                # draw two rows
+                row1 = "........"
+                row2 = "........"
+                row3 = "........"
+                for i in range(24, self.patternLength):
+                    row4 = row4 + "."
+            elif self.patternLength > 16:
+                row1 = "........"
+                row2 = "........"
+                for i in range(16, self.patternLength):
+                    row3 = row3 + "."
+            elif self.patternLength > 8:
+                row1 = "........"
+                for i in range(8, self.patternLength):
+                    row2 = row2 + "."
+            else:
+                # draw one row
+                for i in range(self.patternLength):
+                    row1 = row1 + "."
+
+            xStart = 27
+            oled.text(row1, xStart, 0, 1)
+            oled.text(row2, xStart, 6, 1)
+            oled.text(row3, xStart, 12, 1)
+            oled.text(row4, xStart, 18, 1)
+
+        # Draw a visual cue for when a long button press has been detected
+        # and a new random pattern is being generated
+        if self.shreadedVis:
+            fb = framebuf.FrameBuffer(bytearray(b'\x0f\x000\x80N`Q \x94\xa0\xaa\x90\xa9P\xa5@Z\x80H\x803\x00\x0c\x00'), 12, 12, framebuf.MONO_HLSB)
+            oled.blit(fb, 0,0)
+
+        oled.show()
+
+    # -----------------------------
+    # Slew functions
+    # -----------------------------
+
+
+    def stepUpStepDown(self, start, stop, num, buffer):
+        """Produces step up, step down"""
+        c = 0
+        if self.patternLength == 1:  # LFO Mode, make sure we complete a full cycle
+            for i in range(num / 2):
+                buffer[c] = start
+                c += 1
+            for i in range(num / 2):
+                buffer[c] = stop
+                c += 1
+        else:
+            for i in range(num - 1):
+                buffer[c] = stop
+                c += 1
+        return buffer

```

Comment:
[question]
[required]
This applies to this function, and all the similar ones below:

1) Please summarize the significance of the parameters in the docstring
2) `buffer` is passed in as a parameter, but it's also the return value. Is the return ever actually used?
3) If you're having memory issues, do you need to keep the values buffered, or can you refactor the code to calculate the needed voltage for the given clock tick & wave shape dynamically?  (I had similar memory issues with Pam's, and wound up rewriting it to dynamically generate the sin/square/triangle/etc... wave shapes, rather than keeping the shape buffered in an array).

Reply:
The use of `buffer` is me attempting to use C-like pointers to keep memory usage to a minimum.
I agree it would be more memory-efficient to refactor the functions to return one sample at a time. However, I would like to avoid this level of refactoring unless it is absolutely necessary which I don't think it is.

Reply:
updated docstring